tests/test_android_ftrace.py

```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env python3
+# SPDX-License-Identifier: MIT
+"""
+Module to test reading ftrace from android devices
+"""
+
+import time
+
+from typing import List, Mapping
+from benchkit.adb import AndroidDebugBridge
+
+
+TRACE_BUFFER_SIZE_KB: int = 96000 
+DEVICE_DUMP_PATH: str = "/sdcard/tmp/ftrace_dump"
+HOST_DUMP_PATH: str = "./tests/tmp/ftrace_dump"
+EXAMPLE_WEBSITE: str = "https://github.com/open-s4c/benchkit"
+HOST_SLEEP_TIME: float = 5
+
+
+def enable_tracing(bridge: AndroidDebugBridge) -> None:
+    # per category tracing
+    bridge.shell_out("echo 1 > /sys/kernel/tracing/events/irq/enable")
+    # per event tracing
+    bridge.shell_out("echo 1 > /sys/kernel/tracing/events/sched/sched_wakeup/enable")
+
+
+def disable_tracing(bridge: AndroidDebugBridge) -> None:
+    bridge.shell_out("echo 0 > /sys/kernel/tracing/events/irq/enable")
+    bridge.shell_out("echo 0 > /sys/kernel/tracing/events/sched/sched_wakeup/enable")
+
+
+def start_tracing(bridge: AndroidDebugBridge) -> None:
+    bridge.shell_out("echo 1 > /sys/kernel/tracing/tracing_on")
+
+
+def stop_tracing(bridge: AndroidDebugBridge) -> None:
+    bridge.shell_out("echo 0 > /sys/kernel/tracing/tracing_on")
+
+
+def dump_trace(bridge: AndroidDebugBridge, output: str) -> None:
+    bridge.shell_out(f"rm {output} || true") # delete file or ignore rm if fails (file not exists)
+    bridge.shell_out(f"cat /sys/kernel/tracing/trace > {output}")
+
+
+def clear_trace_buffer(bridge: AndroidDebugBridge) -> None:
+    bridge.shell_out("cat /dev/null > /sys/kernel/tracing/trace")
+
+
+def set_trace_buffer_size(bridge: AndroidDebugBridge, size_kb: int) -> None:
+    bridge.shell_out(f"echo {size_kb} > /sys/kernel/tracing/buffer_size_kb")
+
+
+def open_website(bridge: AndroidDebugBridge, url: str) -> None:
+    bridge.shell_out(f"am start -a android.intent.action.VIEW -d \"{url}\" com.android.chrome")
+
+
+def collect_spans(file: str) -> List[Mapping]:
+    events = []
+    event_stack = []
+    
+    # TODO: fix this parsing, it seems to be incorrect
+    with open(file, "r") as f:
```

Comment:
We should be able to use the perfetto trace processor. https://perfetto.dev/docs/analysis/trace-processor-python
Did you try it out already and ran into issues?

Reply:
no i did not, but this looks like a very promising idea, I'll try it out.